src/librustc_codegen_ssa/mir/operand.rs

```diff
@@ -380,45 +380,51 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
+            // alloca; they are handled somewhat differently
+            if let mir::PlaceBase::Local(index) = place_base {
+                match self.locals[*index] {
+                    LocalRef::Operand(mut opt @ Some(_)) => {
+                        // Moves out of scalar and scalar pair fields are trivial.
+                        for proj in place_projection {
+                            if let Some(o) = opt {
+                                match proj.elem {
+                                    mir::ProjectionElem::Field(ref f, _) => {
+                                        opt = Some(o.extract_field(bx, f.index()));
+                                        continue;
+                                    }
+                                    mir::ProjectionElem::Index(_) |
+                                    mir::ProjectionElem::ConstantIndex { .. } => {
+                                        // ZSTs don't require any actual memory access.
+                                        // FIXME(eddyb) deduplicate this with the identical
+                                        // checks in `codegen_consume` and `extract_field`.
+                                        let elem = o.layout.field(bx.cx(), 0);
+                                        if elem.is_zst() {
+                                            opt = Some(OperandRef::new_zst(bx, elem));
+                                            continue;
+                                        }
+                                    }
+                                    _ => {}
+                                }
+
+                                return None;
+                            }
+                        }
 
-        // Moves out of scalar and scalar pair fields are trivial.
-        if let &mir::Place::Projection(ref proj) = place {
-            if let Some(o) = self.maybe_codegen_consume_direct(bx, &proj.base) {
-                match proj.elem {
-                    mir::ProjectionElem::Field(ref f, _) => {
-                        return Some(o.extract_field(bx, f.index()));
+                        return opt;
                     }
-                    mir::ProjectionElem::Index(_) |
-                    mir::ProjectionElem::ConstantIndex { .. } => {
-                        // ZSTs don't require any actual memory access.
-                        // FIXME(eddyb) deduplicate this with the identical
-                        // checks in `codegen_consume` and `extract_field`.
-                        let elem = o.layout.field(bx.cx(), 0);
-                        if elem.is_zst() {
-                            return Some(OperandRef::new_zst(bx, elem));
-                        }
+                    LocalRef::Operand(None) => {
+                        bug!("use of {:?} before def", place);
+                    }
+                    LocalRef::Place(..) | LocalRef::UnsizedPlace(..) => {
+                        // use path below
```

Comment:
Looking at the old code, this seems to make no sense to me. It used to just return `None`. If https://github.com/rust-lang/rust/blob/f5712d2de09e2c35843150b05fdf3672534dff00/src/librustc_codegen_ssa/mir/operand.rs#L385 is matched, then https://github.com/rust-lang/rust/blob/f5712d2de09e2c35843150b05fdf3672534dff00/src/librustc_codegen_ssa/mir/operand.rs#L400 can't possibly match. I therefor think we should move the 

```
            // watch out for locals that do not have an
            // alloca; they are handled somewhat differently
```

comment from above to here and just do `None` here. This also means we can make the `return opt;` just be `opt` (or with my suggested changes above: `Some(o)`